### PR TITLE
Add endpoint detection for streaming greedy search

### PR DESCRIPTION
**Pull request by csukuangfj**

### CODE REVIEW

**Comment by shaynemei** [comment]
Sure I can have a look next Tuesday and test it out. I’m away from keyboard this weekend.

**Comment by csukuangfj** [comment]
> Sure I can have a look next Tuesday and test it out. I’m away from keyboard this weekend.

Thanks

**Comment by ezerhouni** [comment]
@csukuangfj Thank you for taking care of it. I will try to have a look also beginning of next week

**Comment by csukuangfj** [comment]
@ezerhouni 

Thanks!

File: sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py [hunk 1]
```diff
@@ -483,8 +507,19 @@ async def handle_connection_impl(
...
+                if stream.endpoint_detected(self.online_endpoint_config):
+                    final = 1
+
+                message = {
+                    "text": hyp,
+                    "final": final,
+                }
+
+                await socket.send(json.dumps(message))
+                if final:
```
Comment (shaynemei) [inline sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py:521]: Maybe we shouldn't return here. I thought the goal of an endpointer is to provide additional information to show sth like VAD segments/pauses in continuous speech? If we return here, the connection would close on the first segment.

Comment (danpovey) [reply sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py:521]: This makes sense I think.  the way it was done in Kaldi was for demonstration purposes, really.

Comment (shaynemei) [reply sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py:521]: I see. 
In terms of the endpointer functionality, after slightly modifying it to not return on the first segment and resetting `stream.hyp` as follows:
snap 2022-09-07 at 11 46 27 PM

the endpointing detection works fine in the beginning but seems to start showing unexpected behaviour at some point, as you can see towards the bottom of the screenshot, many endpoints were detected:

snap 2022-09-07 at 11 43 45 PM

Maybe there's a bug somewhere in the endpointer?

Comment (csukuangfj) [reply sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py:521]: There are three default rules in endpoint detection. If any of them is activated, then we assume endpoint is detected.

I suspect that rule 3 is triggered.

You can find the 3 rules in
https://github.com/k2-fsa/sherpa/blob/9e5bc4ec7ff6fffb95b932cb092b4209932ad20f/sherpa/python/sherpa/online_endpoint.py#L29-L49

---

We provide the following commandline arguments to change them:
https://github.com/k2-fsa/sherpa/blob/9e5bc4ec7ff6fffb95b932cb092b4209932ad20f/sherpa/python/test/test_online_endpoint.py#L116-L118

---

You can pass `--help` to see the usage.

Comment (shaynemei) [reply sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py:521]: I see, thanks! I have no further questions.

**Review by shaynemei — APPROVED** [review]
LGTM

**Comment by ezerhouni** [comment]
@csukuangfj I have been a been a bit busy lately. Can I try it out this branch/do a quick review or should I wait for other commits (based on previous review) ? Thank you

**Comment by csukuangfj** [comment]
> @csukuangfj I have been a been a bit busy lately. Can I try it out this branch/do a quick review or should I wait for other commits (based on previous review) ? Thank you

Could you wait for a moment? I am making changes to it.

**Comment by ezerhouni** [comment]
> > @csukuangfj I have been a been a bit busy lately. Can I try it out this branch/do a quick review or should I wait for other commits (based on previous review) ? Thank you
> 
> Could you wait for a moment? I am making changes to it.

Sure ! Didn't wanted to stress you out ! Take your time

**Comment by danpovey** [comment]
It would be great if @ezerhouni could help with the review!

File: sherpa/bin/conv_emformer_transducer_stateless2/streaming_server.py [hunk 2]
```diff
@@ -426,10 +445,14 @@ async def handle_connection_impl(
         stream = Stream(
             context_size=self.context_size,
             initial_states=self.initial_states,
+            subsampling_factor=self.subsampling_factor,
         )
 
         self.beam_search.init_stream(stream)
 
+        # If set, it means some endpointing rule is activated.
+        final = 0
```
Comment (ezerhouni) [inline sherpa/bin/conv_emformer_transducer_stateless2/streaming_server.py:454]: Nitpicking : ```endpoint_detected``` returns a ```bool```. What do you think about using ```bool``` for ```final``` also ?

File: sherpa/bin/pruned_stateless_emformer_rnnt2/streaming_server.py [hunk 3]
```diff
@@ -447,7 +465,23 @@ async def send_results():
...
+                    final = 0
+
+                message = {
+                    "segment": stream.segment,
+                    "text": hyp,
+                    "final": final,
+                }
+
+                await socket.send(json.dumps(message))
+                if final:
```
Comment (ezerhouni) [inline sherpa/bin/pruned_stateless_emformer_rnnt2/streaming_server.py:483]: This condition can be added in L470 isn't it ?

File: sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py [hunk 4]
```diff
@@ -483,8 +507,23 @@ async def handle_connection_impl(
 
             while len(stream.features) > self.chunk_length:
                 await self.compute_and_decode(stream)
-                result = self.beam_search.get_texts(stream)
-                await socket.send(result)
+                hyp = self.beam_search.get_texts(stream)
+
+                if stream.endpoint_detected(self.online_endpoint_config):
```
Comment (ezerhouni) [inline sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py:512]: Maybe something like : 
```
is_final = stream.endpoint_detected(self.online_endpoint_config)
if is_final:
    self.beam_search.init_stream(stream)
    stream.segment += 1

message = {"segment": stream.segment - 1,
                     "text": hyp,
                    "final": is_final}
```

**Review by ezerhouni — COMMENTED** [review]
Few (nitpicking) comments, otherwise looks good to me @csukuangfj 
Thank you very much for taking care of it. I will test the code this afternoon and let you know if something wrong pops up.

File: sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py (hunk 4 above)
Comment (csukuangfj) [reply sherpa/bin/streaming_pruned_transducer_statelessX/streaming_server.py]: Thanks! That sounds good. I just made the changes in the latest commit.

File: sherpa/bin/conv_emformer_transducer_stateless2/streaming_server.py [hunk 5]
```diff
@@ -466,21 +463,18 @@ async def handle_connection_impl(
                 await self.compute_and_decode(stream)
                 hyp = self.beam_search.get_texts(stream)
 
-                if stream.endpoint_detected(self.online_endpoint_config):
-                    final = 1
+                segment = stream.segment
+                is_final = stream.endpoint_detected(self.online_endpoint_config)
+                if is_final:
```
Comment (ezerhouni) [inline sherpa/bin/conv_emformer_transducer_stateless2/streaming_server.py:468]: @csukuangfj I think you removed the ``` segment + 1```

Comment (ezerhouni) [reply sherpa/bin/conv_emformer_transducer_stateless2/streaming_server.py:468]: Ah my bad I mis-red the code. It should be fine

**Review by ezerhouni — APPROVED** [review]

**Pull request merged**